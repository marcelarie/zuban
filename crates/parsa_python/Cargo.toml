[package]
name = "parsa_python"
version.workspace = true
edition.workspace = true
rust-version.workspace = true
publish = false
homepage.workspace = true
authors.workspace = true

[dependencies]
parsa.workspace = true
lazy_static.workspace = true
regex.workspace = true

[dev-dependencies]
insta.workspace = true
test_utils.workspace = true
