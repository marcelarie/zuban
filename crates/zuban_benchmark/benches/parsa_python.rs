use criterion::{criterion_group, criterion_main, Criterion};

use parsa_python_cst::Tree;

fn parse_file(copies: usize) {
    let some_code = utils::dedent(
        r#"
        import os
        def x(a, b, c):
            try:
                return a
            except Exception:
                return [1, 2, 3, 4, 5, 6]
        class C:
            x: int
            y = 1 # type: int
            z: dict[int, list[str]] = {1: ["", r''], 2: [f""" {1} """]}
            def f(self, x: list[str], y: int): ...
    "#,
    );
    assert_eq!(some_code.len(), 262);
    Tree::parse(some_code.repeat(copies).into());
}

fn bench_parser(c: &mut Criterion) {
    c.bench_function("parse big file", |b| b.iter(|| parse_file(1000)));
    c.bench_function("parse small file", |b| b.iter(|| parse_file(1)));
}

// Register the benchmarks
criterion_group!(benches, bench_parser);
criterion_main!(benches);
