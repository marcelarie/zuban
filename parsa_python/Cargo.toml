[package]
name = "parsa_python"
version = "0.1.0"
edition.workspace = true
rust-version.workspace = true
homepage.workspace = true
authors.workspace = true

[dependencies]
parsa.workspace = true
lazy_static.workspace = true
regex.workspace = true

[dev-dependencies]
insta.workspace = true
