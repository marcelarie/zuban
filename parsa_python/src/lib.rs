use parsa::{create_parser,create_node,create_token};


create_token!(PythonToken, enum TokenType, String, Number, Endmarker);
create_node!(PythonNode, enum NodeType, file_input, foo);

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn it_works() {
        assert_eq!(2 + 2, 4);
        parse_python("bar");
        dbg!(TokenType::get_map());
        dbg!(NodeType::get_map());
        return
    }
}

struct PythonTokenizer {
}

impl PythonTokenizer {
    fn new_tok(&self) -> PythonToken {
        PythonToken {
            start: 0,
            length: 0,
            type_: TokenType::Endmarker,
        }
    }
}

impl parsa::Tokenizer<PythonToken> for PythonTokenizer {
    fn new(_code: &str) -> Self {
        Self {}
    }

    fn yield_next(&self) -> PythonToken {
        self.new_tok()
    }

}

create_parser!(parse_python, PythonTree, PythonNode,
               PythonToken, PythonTokenizer, TokenType, NodeType);
//create_parser!(parse_pythonx, PythonTreex, PythonNodex,
//               PythonToken, PythonTokenizer, TokenType, NodeType);
