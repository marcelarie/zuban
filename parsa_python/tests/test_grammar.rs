extern crate insta;

use parsa_python::*;

fn dedent(s: &'static str) -> String {
    let mut indent = usize::MAX;
    let lines: &Vec<_> = &s.split('\n').collect();
    for line in lines {
        if !line.trim().is_empty() {
            indent = std::cmp::min(indent, line.len() - line.trim_start().len());
        }
    }
    if indent == usize::MAX {
        return s.to_string();
    }
    let new_lines: Vec<_> = lines
        .iter()
        .map(|line| {
            if line.len() >= indent {
                &line[indent..]
            } else {
                line
            }
        })
        .collect();
    new_lines.join("\n")
}

fn tree_to_string(tree: PyTree) -> String {
    fn recurse(code: &mut String, node: &PyNode, depth: usize) {
        *code += &" ".repeat(depth);
        *code += &format!(
            "{}: {}-{}{}\n",
            node.type_str(),
            node.start(),
            node.end(),
            if node.is_leaf() {
                format!(" {:?}", node.as_code())
            } else {
                "".to_string()
            }
        );
        for c in node.iter_children() {
            assert_eq!(node.index, c.parent().unwrap().index);
            recurse(code, &c, depth + 1);
        }
    }

    let root = tree.root_node();
    assert!(root.parent().is_none());
    let mut code = String::new();
    recurse(&mut code, &root, 0);
    code
}

macro_rules! parametrize_snapshots {
    ($($name:ident : $input:expr;)*) => {$(
        #[test]
        fn $name() {
            let tree = parse($input.into());
            insta::assert_snapshot!(stringify!($name), tree_to_string(tree));
        }
    )*}
}

parametrize_snapshots!(
    empty: "".to_owned();
    very_simple: "1".to_owned();
    json: "{foo: 1}\n".to_owned();
    simple: dedent("
        a(bar:=1, foo=1);
        if a:
         1
        else: 2
        ");
    assignments_simple: dedent("
        a = 1
        a, b = 2
        a, b, = 2
        (a, b) = 2
        ((a, b)) = 2
        a = b = 3
        a = (b, c) = 3
        a = (b, c,) = 3
        a = b, c, = 3
        ");
    assignments_star: dedent("
        *foo = *bar = *baz
        i, *foo = ((j, [*bar, k])) = *baz
        *foo, *bar = yield baz
        ");
    assignments_aug: dedent("
        foo += 3
        foo.bar |= 3
        foo, bar += 3
        ");
    assignments_setitem: dedent("
        foo[0] += 3
        foo[0] = a = 3
        foo[a, b] = c
        foo[a::, b:f, c:=3] = c
        ");
    assignments_annotation: dedent("
        foo: bar
        foo: int = 3
        foo.bar: baz
        foo: bar = a = 4
        ");
    assignments_with_call: dedent("
        foo().bar = 1
        foo(1,).bar[1] = 1
        foo(a for a in b).bar[1] = 1
        ");
    for_stmt: dedent("
        for x in [1,2]:
            ...
        for *b in *c: ...
        for c.x, (*b, d.y[0]) in i, *j, k: ...
        ");
    comprehension_simple: dedent("
        [1 for a in b]
        [a for *a, c in b if c if d]
        foo(a for (*a, (*c, b),), in d)
        ");
    comprehension_fails: dedent("
        # fails
        [1 for a in *b]
        [1 for a in b if a else c]
        ");
    del_stmt: dedent("
        del foo
        del foo,
        del foo, bar.baz
        del foo[1]
        ");
    cls: dedent("
        class Foo(object):
            def __init__(self, /, f, *, g):
                pass

            @property
            def prop(self):
                pass
        ");
    import_names: dedent("
        import foo
        import foo.bar as baz
        import foo, bar, baz
        import foo.bar.baz
        ");
    import_name_fails: dedent("
        import foo,
        import foo, bar,
        ");
    import_from_normal: dedent("
        from foo import bar
        from foo import bar as baz
        from foo.bar.baz import *
        ");
    import_from_dotted: dedent("
        from . import (foo as baz, bar)
        from . import (foo, bar,)
        from ..foo import bar as baz
        from ... import *
        ");
    import_from_fails: dedent("
        from . import foo,
        from . import foo, bar,
        ");
    calls: dedent("
        foo(bar)
        foo(bar_foo=3)
        foo(bar_baz:=2)
        foo(**kwargs)
        ");
    nested_call: dedent("
        foo(1)()
        ");
    complex_calls: dedent("
        foo(bar, 3)
        foo(bar, *foo)
        foo(bar, *foo, name=3)
        foo(bar, *foo, name=3, **kwargs)
        foo(bar, *foo, name=3, **kwargs, b=3)
        foo(**kwargs, b=3, **foo)
        ");
    operators: dedent("
        f | d | c + 5 & 2 * 3 * 4 ** 5 ** 6
        ");
    simple_error_recovery: dedent("
        a + 3 +
        b = 3
        ");
    terminal_error_recovery: dedent("
        ?
        ");
    terminal_and_nonterminal_error_recovery: dedent("
        foo
        (1) + ? + c
        bar
        ");
    keyword_recovery: dedent("
        foo
        def
        bar
        else:
        baz
        ");
    match_simple: dedent("
        match foo:
            case ['a']: ...
            case _:
                pass
        ");
    soft_keyword_underscore: dedent("
        match _:
            case ['a']: _(_)
            case _:
                _
        match = 3
        ");
    match_underscore_with_as: dedent("
        match foo:
            case bar as _: pass
        ");
    match_class_pattern: dedent("
        match a:
            case A():
                pass
            case B(1, 2):
                pass
            case B(1, b=2):
                pass
            case B(a=1, b=2):
                pass
        ");
    dict_literal1: dedent("
        {1: 2}
        {**foo}
        {**foo, 1: 2}
        {**foo, **bar}
        ");
    function_simple: dedent("
        def foo(): ...
        def foo(a: 3): ...
        def foo(a, b,): ...
        def foo(a: f = 3, b=4): ...
        def foo(a=3, b=4): ...
        ");
    function_args: dedent("
        def foo(a, *, b=3): ...
        def foo(a=3, *, c): ...
        def foo(a, *, x, **kwargs): ...
        def foo(a, *args: int, **fooo: str,): ...
        def foo(a, /, b: i, c: i=3, **kwargs): ...
        def foo(*args): ...
        ");
    lambda_simple: dedent("
        lambda: ...
        lambda a: ...
        lambda a, b,: ...
        lambda a=3, b=4: ...
        ");
    lambda_slash: dedent("
        lambda a, /, b=3: ...
        lambda a=3, /,: ...
        lambda a=3, /: ...
        lambda a=3, /, *args, c=3: ...
        lambda a=3, /, c=3: ...
        lambda a, /, b, c=3, **kwargs: ...
        lambda a, a=3, /, c=3, *, d: ...
        ");
    lambda_args: dedent("
        lambda a, *, b=3: ...
        lambda a=3, *, c: ...
        lambda a, *, x, **kwargs: ...
        lambda a, *args, **fooo ,: ...
        lambda *args: ...
        ");
    lambda_kwargs: dedent("
        lambda **kwargs: ...
        lambda a=3, **kwargs: ...
        lambda a, **kwargs: ...
        lambda a, /, b, *, x, **kwargs: ...
        ");
    lambda_fails: dedent("
        lambda a, *, **kwargs: ...
        lambda a=3, b: ...
        lambda **a, b: ...
        lambda a=3, *,: ...
        lambda ,: ...
        ");
    lambda_fail_with_slash: dedent("
        lambda a=3, /, b: ...
        lambda a=3, /, b=3, c: ...
        lambda a, /, b=3, c: ...
        lambda z, a=3, d, /, b: ...
        ");
    with_items_simple: dedent("
        with foo as bar: ...
        with foo as bar, baz: ...
        ");
    with_items_parentheses: dedent("
        with (foo as bar): ...
        with (foo as bar, baz): ...
        with ((foo)): ...
        with (i for i in 42): ...
        with (yield y): ...
        ");
    string_literals_in_strings: dedent(r#"
        '"""'
        "'''"
        ""
        ''
        "#);
    bytes: dedent(r#"
        b'asdf\n\''
        br'\asdf'
        Br'asdf"""'
        bR''
        BR'\''
        "#);
    exception_group: dedent(r#"
        try:
            ...
        except* SpamError:
            ...
        except* FooError as e:
            ...
        except* (BarError, BazError) as e:
            ...
        "#);
    invalid_syntax: dedent(r#"
        def f():
            (x,) += 1
            z
        "#);
);
