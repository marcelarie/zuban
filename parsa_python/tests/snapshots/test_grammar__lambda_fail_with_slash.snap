---
source: parsa_python/tests/test_grammar.rs
expression: tree_to_string(tree)
---
file: 0-103
 Error(stmt): 1-17
  Error(simple_stmts): 1-17
   Error(simple_stmt): 1-17
    Error(star_expressions): 1-17
     Error(expression): 1-17
      Error(lambda): 1-17
       Keyword: 1-7 "lambda"
       Error(lambda_parameters): 8-17
        lambda_param_with_default: 8-11
         name_definition: 8-9
          Name: 8-9 "a"
         Keyword: 9-10 "="
         expression: 10-11
          atom: 10-11
           Number: 10-11 "3"
        Keyword: 11-12 ","
        Keyword: 13-14 "/"
        Keyword: 14-15 ","
        Error(lambda_param_with_default): 16-17
         name_definition: 16-17
          Name: 16-17 "b"
 Error(stmt): 17-18
  Error(Keyword): 17-18 ":"
 stmt: 19-23
  simple_stmts: 19-23
   simple_stmt: 19-22
    star_expressions: 19-22
     expression: 19-22
      atom: 19-22
       Keyword: 19-22 "..."
   Newline: 22-23 "\n"
 Error(stmt): 23-44
  Error(simple_stmts): 23-44
   Error(simple_stmt): 23-44
    Error(star_expressions): 23-44
     Error(expression): 23-44
      Error(lambda): 23-44
       Keyword: 23-29 "lambda"
       Error(lambda_parameters): 30-44
        lambda_param_with_default: 30-33
         name_definition: 30-31
          Name: 30-31 "a"
         Keyword: 31-32 "="
         expression: 32-33
          atom: 32-33
           Number: 32-33 "3"
        Keyword: 33-34 ","
        Keyword: 35-36 "/"
        Keyword: 36-37 ","
        lambda_param_with_default: 38-41
         name_definition: 38-39
          Name: 38-39 "b"
         Keyword: 39-40 "="
         expression: 40-41
          atom: 40-41
           Number: 40-41 "3"
        Keyword: 41-42 ","
        Error(lambda_param_with_default): 43-44
         name_definition: 43-44
          Name: 43-44 "c"
 Error(stmt): 44-45
  Error(Keyword): 44-45 ":"
 stmt: 46-50
  simple_stmts: 46-50
   simple_stmt: 46-49
    star_expressions: 46-49
     expression: 46-49
      atom: 46-49
       Keyword: 46-49 "..."
   Newline: 49-50 "\n"
 Error(stmt): 50-58
  Error(simple_stmts): 50-58
   Error(simple_stmt): 50-58
    Error(star_expressions): 50-58
     Error(expression): 50-58
      Error(lambda): 50-58
       Keyword: 50-56 "lambda"
       Error(lambda_parameters): 57-58
        Error(lambda_param_with_default): 57-58
         name_definition: 57-58
          Name: 57-58 "a"
 Error(stmt): 58-59
  Error(Keyword): 58-59 ","
 Error(stmt): 60-61
  Error(Keyword): 60-61 "/"
 Error(stmt): 61-62
  Error(Keyword): 61-62 ","
 Error(stmt): 63-69
  Error(simple_stmts): 63-69
   simple_stmt: 63-69
    assignment: 63-69
     star_targets: 63-64
      star_target: 63-64
       target_with_star_atom: 63-64
        star_atom: 63-64
         name_definition: 63-64
          Name: 63-64 "b"
     Keyword: 64-65 "="
     star_expressions: 65-69
      expression: 65-66
       atom: 65-66
        Number: 65-66 "3"
      Keyword: 66-67 ","
      expression: 68-69
       atom: 68-69
        Name: 68-69 "c"
 Error(stmt): 69-70
  Error(Keyword): 69-70 ":"
 stmt: 71-75
  simple_stmts: 71-75
   simple_stmt: 71-74
    star_expressions: 71-74
     expression: 71-74
      atom: 71-74
       Keyword: 71-74 "..."
   Newline: 74-75 "\n"
 Error(stmt): 75-83
  Error(simple_stmts): 75-83
   Error(simple_stmt): 75-83
    Error(star_expressions): 75-83
     Error(expression): 75-83
      Error(lambda): 75-83
       Keyword: 75-81 "lambda"
       Error(lambda_parameters): 82-83
        Error(lambda_param_with_default): 82-83
         name_definition: 82-83
          Name: 82-83 "z"
 Error(stmt): 83-84
  Error(Keyword): 83-84 ","
 Error(stmt): 85-92
  Error(simple_stmts): 85-92
   simple_stmt: 85-92
    assignment: 85-92
     star_targets: 85-86
      star_target: 85-86
       target_with_star_atom: 85-86
        star_atom: 85-86
         name_definition: 85-86
          Name: 85-86 "a"
     Keyword: 86-87 "="
     star_expressions: 87-92
      expression: 87-88
       atom: 87-88
        Number: 87-88 "3"
      Keyword: 88-89 ","
      expression: 90-91
       atom: 90-91
        Name: 90-91 "d"
      Keyword: 91-92 ","
 Error(stmt): 93-94
  Error(Keyword): 93-94 "/"
 Error(stmt): 94-95
  Error(Keyword): 94-95 ","
 stmt: 96-103
  simple_stmts: 96-103
   simple_stmt: 96-102
    assignment: 96-102
     single_target: 96-97
      name_definition: 96-97
       Name: 96-97 "b"
     Keyword: 97-98 ":"
     expression: 99-102
      atom: 99-102
       Keyword: 99-102 "..."
   Newline: 102-103 "\n"
 Endmarker: 103-103 ""

